encode.go

```diff
@@ -152,163 +118,159 @@ func encodeUint64(val reflect.Value, w *encbuf) error {
+					return fmt.Errorf("failed to encode field of struct: %v", err)
+				}
+				variableParts = append(variableParts, elemBuf.str)
+				fixedParts = append(fixedParts, []byte{})
 			}
 		}
-		totalSize := len(w.str) - lengthBytes - origBufSize
-		if totalSize >= 2<<32 {
-			return errors.New("struct oversize")
+		serializedStruct, err := serializeFromParts(fixedParts, variableParts, len(fields))
+		if err != nil {
+			return err
 		}
-		binary.LittleEndian.PutUint32(totalSizeEnc, uint32(totalSize))
-		copy(w.str[origBufSize:origBufSize+lengthBytes], totalSizeEnc)
+		w.str = append(w.str, serializedStruct...)
 		return nil
 	}
-	encodeSizer := func(val reflect.Value) (uint32, error) {
-		totalSize := uint32(0)
-		for _, f := range fields {
-			fieldSize, err := f.sszUtils.encodeSizer(val.Field(f.index))
-			if err != nil {
-				return 0, fmt.Errorf("failed to get encode size for field of struct: %v", err)
-			}
-			totalSize += fieldSize
-		}
-		return lengthBytes + totalSize, nil
-	}
-	return encoder, encodeSizer, nil
+	return encoder, nil
 }
 
-func makePtrEncoder(typ reflect.Type) (encoder, encodeSizer, error) {
+func makePtrEncoder(typ reflect.Type) (encoder, error) {
 	elemSSZUtils, err := cachedSSZUtilsNoAcquireLock(typ.Elem())
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-
-	// After considered the use case in Prysm, we've decided that:
-	// - We assume we will only encode/decode pointer of array, slice or struct.
-	// - The encoding for nil pointer shall be 0x00000000.
 	encoder := func(val reflect.Value, w *encbuf) error {
+		// Nil encodes to []byte{}.
 		if val.IsNil() {
-			totalSizeEnc := make([]byte, lengthBytes)
-			w.str = append(w.str, totalSizeEnc...)
+			w.str = append(w.str, []byte{}...)
 			return nil
 		}
 		return elemSSZUtils.encoder(val.Elem(), w)
 	}
 
-	encodeSizer := func(val reflect.Value) (uint32, error) {
-		if val.IsNil() {
-			return lengthBytes, nil
-		}
-		return elemSSZUtils.encodeSizer(val.Elem())
-	}
-
-	return encoder, encodeSizer, nil
+	return encoder, nil
 }
 
-// encodeError is what gets reported to the encoder user in error case.
-type encodeError struct {
-	msg string
-	typ reflect.Type
-}
-
-func newEncodeError(msg string, typ reflect.Type) *encodeError {
-	return &encodeError{msg, typ}
+func serializeFromParts(fixedParts [][]byte, variableParts [][]byte, numElements int) ([]byte, error) {
+	fixedLengths := []int{}
+	variableLengths := []int{}
+	for _, item := range fixedParts {
+		if !bytes.Equal(item, []byte{}) {
+			fixedLengths = append(fixedLengths, len(item))
+		} else {
+			fixedLengths = append(fixedLengths, BytesPerLengthOffset)
+		}
+	}
+	for _, item := range variableParts {
+		if len(item) == 0 {
+			continue
+		}
+		variableLengths = append(variableLengths, len(item))
+	}
+	sum := 0
+	for _, item := range append(fixedLengths, variableLengths...) {
+		sum += item
+	}
+	if sum >= 1<<uint64(BytesPerLengthOffset*BitsPerByte) {
```

Comment:
can this `1<<uint64(BytesPerLengthOffset*BitsPerByte)` be another var, its hard to parse the condition on a first look